File: client/lib/src/crypto/secret_box.dart
import 'dart:typed_data';

import 'package:sodium/sodium.dart';

import 'crypto_exception.dart';

/// XSalsa20-Poly1305 symmetric encryption (secretbox) backed by libsodium.
class SecretBoxCrypto {
  SecretBoxCrypto._();

  /// Encrypts [plaintext] with [key] and [nonce].
  ///
  /// [key] must be 32 bytes; [nonce] must be 24 bytes.
  static Uint8List seal(
      Sodium sodium, Uint8List plaintext, Uint8List nonce, Uint8List key) {
    final secKey = SecureKey.fromList(sodium, key);
    try {
      return sodium.crypto.secretBox
          .easy(message: plaintext, nonce: nonce, key: secKey);
    } finally {
      secKey.dispose();
    }
  }

  /// Decrypts [ciphertext] with [key] and [nonce].
  ///
  /// Throws [CryptoException] if authentication fails.
  static Uint8List open(
      Sodium sodium, Uint8List ciphertext, Uint8List nonce, Uint8List key) {
    final secKey = SecureKey.fromList(sodium, key);
    try {
      return sodium.crypto.secretBox
          .openEasy(cipherText: ciphertext, nonce: nonce, key: secKey);
    } catch (e) {
      throw CryptoException('SecretBox decryption failed', e);
    } finally {
      secKey.dispose();
    }
  }
}

File: src/main/java/com/erikromson/datawallet/crypto/SecretBox.java
package com.erikromson.datawallet.crypto;

public final class SecretBox {

    static final int MAC_BYTES = 16;

    private SecretBox() {}

    public static byte[] seal(byte[] plaintext, byte[] nonce, byte[] key) {
        byte[] ciphertext = new byte[plaintext.length + MAC_BYTES];
        int result = CryptoProvider.sodium().crypto_secretbox_easy(
                ciphertext, plaintext, plaintext.length, nonce, key
        );
        if (result != 0) {
            throw new CryptoException("crypto_secretbox_easy failed");
        }
        return ciphertext;
    }

    public static byte[] open(byte[] ciphertext, byte[] nonce, byte[] key) {
        byte[] plaintext = new byte[ciphertext.length - MAC_BYTES];
        int result = CryptoProvider.sodium().crypto_secretbox_open_easy(
                plaintext, ciphertext, ciphertext.length, nonce, key
        );
        if (result != 0) {
            throw new CryptoException("crypto_secretbox_open_easy failed");
        }
        return plaintext;
    }
}
